source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,553 @@
+    FoldConditionBlock(state_.previous_condition_block_, 1);
+  }
+
+  // Only reference to the header block is the backedge in the latch block,
+  // don't change this.
+  state_.new_inst[loop->GetHeaderBlock()->id()] = loop->GetHeaderBlock()->id();
+
+  for (auto& pair : state_.new_blocks) {
+    RemapOperands(pair.second, state_.new_inst);
+  }
+
+  // Swap the state so the new is now the previous.
+  state_.Swap();
+}
+
+uint32_t LoopUnrollerUtilsImpl::GetPhiVariableID(const ir::Instruction* phi,
+                                                 uint32_t label) const {
+  for (uint32_t operand = 3; operand < phi->NumOperands(); operand += 2) {
+    if (phi->GetSingleWordOperand(operand) == label) {
+      return phi->GetSingleWordOperand(operand - 1);
+    }
+  }
+
+  return 0;
+}
+
+void LoopUnrollerUtilsImpl::FoldConditionBlock(ir::BasicBlock* condition_block,
+                                               uint32_t operand_label) {
+  ir::Instruction& old_branch = *condition_block->tail();
+
+  uint32_t new_target = old_branch.GetSingleWordOperand(operand_label);
+  std::unique_ptr<ir::Instruction> new_branch{
+      new ir::Instruction(ir_context_, SpvOp::SpvOpBranch, 0, 0,
+                          {{SPV_OPERAND_TYPE_ID, {new_target}}})};
+  ir_context_->KillInst(&old_branch);
+  condition_block->AddInstruction(std::move(new_branch));
+}
+
+void LoopUnrollerUtilsImpl::CloseUnrolledLoop(ir::Loop* loop) {
+  // Remove the OpLoopMerge instruction from the function.
+  ir::Instruction* merge_inst = loop->GetHeaderBlock()->GetLoopMergeInst();
+  ir_context_->KillInst(merge_inst);
+
+  // Remove the final backedge to the header and make it point instead to the
+  // merge block.
+  state_.previous_continue_block_->tail()->SetInOperand(
+      0, {loop->GetMergeBlock()->id()});
+}
+
+ir::Loop LoopUnrollerUtilsImpl::DuplicateLoop(ir::Loop* old_loop) {
+  const ir::Loop::BasicBlockOrderedListTy& basic_blocks =
+      old_loop->GetOrderedBlocks();
+
+  ir::Loop new_loop = *old_loop;
+
+  ir::Loop::BasicBlockOrderedListTy& new_block_order =
+      new_loop.GetOrderedBlocksRef();
+
+  new_block_order.clear();
+  for (const ir::BasicBlock* itr : basic_blocks) {
+    // Copy the loop basicblock.
+    ir::BasicBlock* BB = itr->Clone(ir_context_);
+    // Assign each result a new unique ID and keep a mapping of the old ids to
+    // the new ones.
+    RemapResultIDs(old_loop, BB);
+
+    if (itr == old_loop->GetLatchBlock()) {
+      new_loop.SetLatchBlock(BB);
+    }
+
+    if (itr == old_loop->GetHeaderBlock()) {
+      new_loop.SetHeaderBlock(BB);
+      new_loop.SetInductionVariable(state_.new_phi);
+    }
+
+    if (itr == old_loop->GetConditionBlock()) {
+      new_loop.SetConditionBlock(BB);
+    }
+
+    blocks_to_add_.push_back(std::unique_ptr<ir::BasicBlock>(BB));
+    state_.new_blocks[itr->id()] = BB;
+    new_block_order.push_back(BB);
+  }
+
+  ir::BasicBlock* new_merge = old_loop->GetMergeBlock()->Clone(ir_context_);
```

Comment:
The merge block might contains other instruction that can only be applied at the end of the loop, they **must not** be cloned.

You need to create a new, empty, merge block that you end up connecting to the correct block. You need to clone and adapt the phi nodes as well.